# HybPhyloMaker support files and scripts

**Support files**  
---
**curcuma_HybSeqProbes_test.fa**  
_test reference containing 100 nuclear exons_  
**Curcuma-roscoeana_plastomeCDS_test.fa**  
_test reference with 11 plastid genes_  

**Scripts**  
---
These scripts are called by main HybPhyloMaker bash scripts.  
Use of these scripts should be properly cited, e.g. as follows:  
  
**alignmentSummary.R**  
_part of HybPhyloMaker_  
**assembled_exons_to_fastas.py**  
https://github.com/listonlab/Hyb-Seq_protocol/blob/master/assembled_exons_to_fastas/assembled_exons_to_fastas.py  
**astral.5.6.1.jar**  
https://github.com/smirarab/ASTRAL  
**catfasta2phyml.pl**  
https://github.com/nylander/catfasta2phyml  
**combineboot.py**  
_part of HybPhyloMaker_ based on http://p4.nhm.ac.uk/tutorial/combine_supports.html  
**CompareToBootstrap.pl**  
http://meta.microbesonline.org/fasttree/treecmp.html  
**CompareTree.pl**  
http://meta.microbesonline.org/fasttree/treecmp.html  
**fastq2fasta.pl**  
http://brianknaus.com/software/srtoolbox/fastq2fasta.pl  
**histogram.r**  
_part of HybPhyloMaker_  
**HybPhyloMaker_reports.R**  
_part of HybPhyloMaker_  
**LBscores.R**  
_part of HybPhyloMaker_ and https://github.com/marekborowiec/metazoan_phylogenomics/blob/master/gene_stats.R  
**MOTree.pm**  
http://meta.microbesonline.org/fasttree/treecmp.html  
**mrp.jar**  
https://github.com/smirarab/mrpmatrix  
**NEBNext-PE.fa**  
Oligonucleotide sequences Â© 2006-2010 Illumina, Inc. All rights reserved.  
**PhiX.fsa**  
http://www.ncbi.nlm.nih.gov/nuccore/9626372  
**plotting_correlations.R**  
_part of HybPhyloMaker_ and https://github.com/marekborowiec/good_genes/blob/master/plotting_correlations.R  
**tree_props.r**  
_part of HybPhyloMaker_ and https://github.com/marekborowiec/good_genes/blob/master/tree_props.R  
**treepropsPlot.r**  
_part of HybPhyloMaker_  
**TreeCollapseCL4.jar**  
http://emmahodcroft.com/TreeCollapseCL.html  
**trimmomatic-0.33.jar**  
http://www.usadellab.org/cms/?page=trimmomatic  
